### PR TITLE
[google_maps_flutter] Fix for memory leak impacting all platforms due to subscriptions not getting cleaned up

### PR DESCRIPTION
**Pull request by wolfie82**
*Prior to this change, GoogleMaps would create a condition where its state wasn't cleaned up.  Any widget that was leveraging GoogleMap wasn't cleaned up if it relied on a subscription, such as onCameraIdle, onCameraMoved, etc.  This was creating crashes for our team specifically.*

*List which issues are fixed by this PR. You must list at least one issue.*

[#129089](https://github.com/flutter/flutter/issues/129089)
[#115283](https://github.com/flutter/flutter/issues/115283)
[#92788](https://github.com/flutter/flutter/issues/92788)

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by wolfie82** [comment]
I'm having a bit of trouble writing a test for this given the public API surface of the internals that were fixed.  @stuartmorgan I hate to open a PR without meeting all the criteria, however, I'm happy to move this forward with a bit of guidance.  I think this PR is extremely important as it fixes a severe memory leak.  In our app, if we navigate away from the GoogleMap and then back again * 3 on an iPhone SE, it will crash with an out of memory exception due to these subscriptions not getting closed and therefore retaining state from the calling widget (see https://github.com/flutter/flutter/issues/129089).

**Comment by stuartmorgan** [comment]
Since the goal is to ensure that the subscriptions are cancelled, it looks like you could test it by vending mock/fake subscriptions in a unit test, and validating that they all have cancel called on them.

**Comment by wolfie82** [comment]
> Since the goal is to ensure that the subscriptions are cancelled, it looks like you could test it by vending mock/fake subscriptions in a unit test, and validating that they all have cancel called on them.

Appreciate that.  This one is a case, at least for me, where the test is harder than the fix.  I'll give it a fresh pair of eyes on Monday.

I was aiming for a test that would leverage the new controller update, but the API surface that's public to the test is what makes this complicated.  I could use the fake platform test implementation already existing, just figuring out how to wire everything up.

**Comment by wolfie82** [comment]
Hi @stuartmorgan.  I added some unit tests and it looks like all the build checks have passed except for `submit-queue`, which I think is outside my control.  Thanks again!

File: packages/google_maps_flutter/google_maps_flutter/test/stream_subscription_test.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/test/stream_subscription_test.dart:1]: This file should be called `controller_test.dart`, since we generally want unit test files to correspond directly to the files they test.

File: packages/google_maps_flutter/google_maps_flutter/lib/src/controller.dart [hunk 1]
```diff
@@ -18,6 +18,35 @@ class GoogleMapController {
   /// The mapId for this controller
   final int mapId;
 
+  /// State map of the stream subscriptions for this controller, used internally.
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/lib/src/controller.dart:21]: Internal implementation details should not be public properties; this would make it impossible for us to ever change without a major version bump.

File: packages/google_maps_flutter/google_maps_flutter/lib/src/controller.dart [hunk 2]
```diff
@@ -18,6 +18,35 @@ class GoogleMapController {
...
+  ///
+  /// The map stores a [StreamSubscription] as key and a [bool] as value.
+  /// Each [StreamSubscription] represents a subscription to an event from the native side.
+  /// The corresponding [bool] indicates whether the subscription has been canceled `true` or not `false`.
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/lib/src/controller.dart:25]: It's unclear to me why the boolean exists, since once it's been canceled there is no reason to track it any more; just being in a set is already a boolean state.

Is this only for the unit test? If so, the unit test should be testing that the subscription is *actually* canceled, via fake subscription that keep track of `cancel` calls, rather than testing an unrelated property that we just hope corresponds to that. As currently written, it would be easy to break the actual functionality without breaking the tests.

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart [hunk 3]
```diff
@@ -398,6 +398,9 @@ class _GoogleMapState extends State<GoogleMap> {
   }
 
   Future<void> onPlatformViewCreated(int id) async {
+    if (!mounted) {
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:401]: How is this related to the PR?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart (hunk 3 above)
Comment (wolfie82) [reply packages/google_maps_flutter/google_maps_flutter/lib/src/google_map.dart:401]: See https://github.com/flutter/flutter/issues/115283 - I can add a test for this.

**Comment by stuartmorgan** [comment]
@wolfie82 Are you still planning on updating this to address the latest feedback?

**Comment by MattyBoy4444** [comment]
Any update on this?

**Comment by stuartmorgan** [comment]
Marking as a draft for now just to get this off of our review queue. Please mark it as ready for review once you've had a chance to update it!

**Comment by Hixie** [comment]
@wolfie82 Thanks for your contribution. I'm going to close this PR for now, but I'll leave a pointer on the issues you listed in case someone wants to take it over. Thanks again!
